File: ch4/c/code4_9.c
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>

int maxSubArraySum(int* arr, int n)
{
  int maxValue = INT_MIN;
  bool allNegative = true;

  // 배열의 값이 모두 음수인 경우, 배열의 값 중 
  // 가장 큰 값을 반환하면 됩니다.
  for(int i = 0; i < n; i++)
  {
    if(arr[i] >= 0)
    {
      allNegative = false;
      break;
    }
    else
    {
      if(arr[i] > maxValue)
        maxValue = arr[i];
    }    
  }
  if(allNegative)
    return maxValue;


  int maxSumSoFar = 0;
  int maxSumEndingHere = 0;

  for(int i = 0; i < n; i++)
  {
    maxSumEndingHere += arr[i];

    if(maxSumEndingHere < 0)
      maxSumEndingHere = 0;

    if(maxSumSoFar < maxSumEndingHere)
      maxSumSoFar = maxSumEndingHere;

    printf("i = %2d, arr[i] = %2d, maxSumEndingHere : %2d, maxSumSoFar : %2d\n",
           i, arr[i], maxSumEndingHere, maxSumSoFar);
  }
  return maxSumSoFar;
}

int main()
{
  int arr[] = {-2, -3, 4, -1, -2, 1, 5, -3};
  int n = sizeof(arr) / sizeof(int);
  printf("부분 배열의 합의 최대값은 %d입니다.\n", maxSubArraySum(arr, n));
  return 0;
}

File: ch4/python/code4_9.py

def max_sub_array_sum(arr, n):
    max_value = -float('inf')
    all_negative = True

    for i in range(0, n):
        if arr[i] >= 0:
            all_negative = False
            break
        elif arr[i] > max_value:
            max_value = arr[i]
    
    if all_negative:
        return max_value
    
    max_sum_so_far = 0
    max_sum_ending_here = 0

    for i in range(0, n):
        max_sum_ending_here += arr[i]

        if max_sum_ending_here < 0:
            max_sum_ending_here = 0
        
        if max_sum_so_far < max_sum_ending_here:
            max_sum_so_far = max_sum_ending_here
        # 표 4-1 확인을 위한 임시 값 출력
        print('i = %d, arr[i] = %2d, maxSumEndingHere : %2d, maxSumSoFar : %2d' % \
          (i, arr[i], max_sum_ending_here, max_sum_so_far))

    return max_sum_so_far

arr = [-2, -3, 4, -1, -2, 1, 5, -3]
n = len(arr)
print('부분 배열의 합의 최댓값은 %d입니다.' % max_sub_array_sum(arr, n))